velox/connectors/hive/HiveDataSink.cpp

```diff
@@ -224,60 +297,91 @@ HiveDataSink::HiveDataSink(
+                       : nullptr) {
+  VELOX_USER_CHECK(
+      !isBucketed() || isPartitioned(),
+      "A bucket table must be partitioned: {}");
+  if (isBucketed()) {
+    VELOX_USER_CHECK_LT(
+        bucketCount_,
+        maxBucketCount(),
+        "bucketCount {} must be smaller than {}",
+        bucketCount_,
+        maxBucketCount());
+  }
+  VELOX_USER_CHECK(
+      (commitStrategy_ == CommitStrategy::kNoCommit) ||
+          (commitStrategy_ == CommitStrategy::kTaskCommit),
+      "Unsupported commit strategy: {}",
+      commitStrategyToString(commitStrategy_));
   // TODO: remove this hack after Prestissimo adds to register dwrf writer.
   facebook::velox::dwrf::registerDwrfWriterFactory();
 }
 
 void HiveDataSink::appendData(RowVectorPtr input) {
   // Write to unpartitioned table.
-  if (partitionChannels_.empty()) {
-    ensureSingleWriter();
-
-    writers_[0]->write(input);
-    writerInfo_[0]->numWrittenRows += input->size();
+  if (!isPartitioned()) {
+    const auto index = ensureWriter(HiveWriterId::unpartitionedId());
+    writers_[index]->write(input);
+    writerInfo_[index]->numWrittenRows += input->size();
     return;
   }
 
   // Write to partitioned table.
-  partitionIdGenerator_->run(input, partitionIds_);
-
-  ensurePartitionWriters();
+  computePartitionAndBucketIds(input);
 
+  // Lazy load all the input columns.
   for (column_index_t i = 0; i < input->childrenSize(); ++i) {
     input->childAt(i)->loadedVector();
   }
 
-  const auto numPartitions = partitionIdGenerator_->numPartitions();
-
   // All inputs belong to a single partition.
-  if (numPartitions == 1) {
-    writers_[0]->write(input);
-    writerInfo_[0]->numWrittenRows += input->size();
+  if (!isBucketed() && partitionIdGenerator_->numPartitions() == 1) {
+    const auto index = ensureWriter(HiveWriterId{0});
+    writers_[index]->write(input);
+    writerInfo_[index]->numWrittenRows += input->size();
     return;
   }
 
-  computePartitionRowCountsAndIndices();
+  splitInputRowsAndEnsureWriters();
 
-  for (auto id = 0; id < numPartitions; id++) {
-    const vector_size_t partitionSize = partitionSizes_[id];
+  for (auto index = 0; index < writers_.size(); ++index) {
+    const vector_size_t partitionSize = partitionSizes_[index];
     if (partitionSize == 0) {
       continue;
     }
 
     RowVectorPtr writerInput = partitionSize == input->size()
         ? input
-        : exec::wrap(partitionSize, partitionRows_[id], input);
-    writers_[id]->write(writerInput);
-    writerInfo_[id]->numWrittenRows += partitionSize;
+        : exec::wrap(partitionSize, partitionRows_[index], input);
+    writers_[index]->write(writerInput);
+    writerInfo_[index]->numWrittenRows += partitionSize;
+  }
+}
+
+void HiveDataSink::computePartitionAndBucketIds(const RowVectorPtr& input) {
+  VELOX_CHECK(isPartitioned());
+  partitionIdGenerator_->run(input, partitionIds_);
+  if (isBucketed()) {
+    bucketFunction_->partition(*input, bucketIds_);
```

Comment:
It is a bit confusing to use "partition" term with bucketing. I don't have any specific suggestion on how to avoid this, but wanted to mention anyway.